### ISSUE TITLE
Strange output of hess_op  in LLSModels

### ISSUE THREAD

**Issue by vepiteski** [JuliaSmoothOptimizers/NLPModels.jl#274]
Consider the following code using NLPModels 0.12.2, LinearOperators 1.0.1 and julia 1.4.0
```
Z=rand(6,3)
v = rand(6)

using NLPModels
using LinearOperators

M = LLSModel(Z,v)

x = rand(3)

H = hess_op(M,v)  #  should be forbidden since the model is of dimension 3 
                  #  but v dimension 6

@show H           # H is nevertheless a 3x3 LinearOperator
@show x

H*x               # resulting in a 6 dimensional vector

```
The execution yields
```
H = Linear operator
  nrow: 3
  ncol: 3
  eltype: Float64
  symmetric: true
  hermitian: true
  nprod:   0
  ntprod:  0
  nctprod: 0


x = [0.6367742195201671, 0.8846556031330619, 0.9993237871283411]
6-element Array{Float64,1}:
 3.5519265702337997
 4.748645449913745
 5.221040330262992
 0.0
 0.0
 0.0
```
so that a 3x3 LinearOperator times a 3-vector gives a 6-vector result.

In julia 1.3.1, NLPModels 0.10.0 and LinearOperators 1.0.1, an error is issued:
```
H = Linear operator
  nrow: 3
  ncol: 3
  eltype: Float64
  symmetric: true
  hermitian: true
  nprod:   0
  ntprod:  0
  nctprod: 0


x = [0.5624829856953892, 0.22147413604981625, 0.5143260460337269]
ERROR: LoadError: DimensionMismatch("array could not be broadcast to match destination")

```
but still H is a 3x3 Operator and x a 3-vector. The issue may well be in LinearOperators, but since the behavior changed from NLPModels 0.10.0 to 0.12.2, I post the issue here.

**Comment by abelsiqueira** [comment]
There are two different errors here.
First, because the output is defined as `Hx = similar(x)`, the output has size 6. The value of `x` is never used, and the dimensions are never checked.
Second, `hprod!` for LLS is defined with `coo_prod!` which also assumes the input is well defined.

@dpo, what do you think? Should we be adding dimension checks to some methods? In some places we decided to accept arrays with larger dimensions. The `Hx = similar(x)` can probably be changed to `Hx = zeros(eltype(x), nlp.meta.nvar)` instead. Still, do we accept `x` with incompatible dimension? And `v`?

**Comment by dpo** [comment]
We've gone back and forth. At some point we said that if the user wants to pass a portion of an existing array or store results in an existing (larger) array, they can always pass views as arguments. So yes, I think we should check dimensions. This isn't Fortran any more.

**Comment by vepiteski** [comment]
Perhaps there is something to validate in LinearOperators as well, we have a LO H confirmed 3x3 multiplied by a 3-vector v giving as result a 6-vector? What is hidden in H that makes H*v 6-dimensionnal? Matrix(H) crashes:
```
julia> Matrix(H)
ERROR: DimensionMismatch("tried to assign 6-element array to 3×1 destination")

```
There is something 6-d hidden in H not visible from outside. I tried a few other functions on H:
```
julia> size(H)
(3, 3)

julia> shape(H)
(3, 3)

julia> diag(H)
ERROR: MethodError: no method matching diag(::LinearOperator{Float64})

julia> check_positive_definite(H)
ERROR: DimensionMismatch("dot product arguments have lengths 3 and 6")

julia> hermitian(H)
true

julia> check_hermitian(H)
ERROR: LinearOperatorException("shape mismatch")

```
diag is in the "latest" documentation but not available in julia1.4.0 LO v1.0.1.

Perhaps all of this discussion should be in LinearOperators but the operator H itself is produced by NLPModels.

**Comment by abelsiqueira** [comment]
Dominique, I'm gonna make updates and tests to check input dimensions.
Jean-Pierre, `H` is defined by default by `v -> hprod(nlp, x, v)`, and the default `hprod` is defined by `hprod!(nlp, x, v, Hx)` with `Hx = similar(x)`, which is the hidden 6-d object. LinearOperators don't check whether the definition makes sense - I don't think it could, in this case - so the problem is here.

**Comment by abelsiqueira** [comment]
This was fixed by #276. The following line returns the error `DimensionError: Input x should have length 3 not 6`:
```julia
H = hess_op(M,v)  #  should be forbidden since the model is of dimension 3 
                  #  but v dimension 6
```
Thanks for reporting it @vepiteski.
